### PR TITLE
Add Dockerfile for fuseml-core

### PR DESCRIPTION
**Pull request by flaviodsr**
Add a dockerfile for building the container image for fuseml-core,
the dockerfile contains 2 stages:
  1. build fuseml-core using golang image
  2. build the fuseml-core container image from docker scratch by simple
    copying the binary built on stage 1 to it and use that as entrypoint.

This change also adds new targets into the makefile for building the
container image and running tests/fuseml_core/fuseml_core-cli.

Resolves: fuseml/fuseml#57

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,37 +1,65 @@
...
+# Generate code
+generate:
 	go mod download
 	go run goa.design/goa/v3/cmd/goa gen github.com/fuseml/fuseml-core/design
 
-example: gen
-	go run goa.design/goa/v3/cmd/goa example github.com/fuseml/fuseml-core/design
-
+# Lint code
 lint: fmt vet tidy
```
Comment (evrardjp) [inline Makefile:38]: We could probably use golint ?

File: cmd/fuseml_core-cli/main.go [hunk 2]
```diff
@@ -99,8 +99,8 @@ func usage() {
 Usage:
     %s [-host HOST][-url URL][-timeout SECONDS][-verbose|-v] SERVICE ENDPOINT [flags]
 
-    -host HOST:  server host (localhost). valid values: localhost
-    -url URL:    specify service URL overriding host URL (http://localhost:8080)
+    -host HOST:  server host (0.0.0.0). valid values: 0.0.0.0
```
Comment (evrardjp) [inline cmd/fuseml_core-cli/main.go:102]: nit/Side note: I would use the code parsing the CLI arguments to generate usage string, instead of having some text that can be easily outdated. cobra isn't a bad library, though a bit big.

File: cmd/fuseml_core-cli/main.go [hunk 3]
```diff
@@ -14,7 +14,7 @@ import (
 
 func main() {
 	var (
-		hostF = flag.String("host", "localhost", "Server host (valid values: localhost)")
+		hostF = flag.String("host", "0.0.0.0", "Server host (valid values: 0.0.0.0)")
```
Comment (evrardjp) [inline cmd/fuseml_core-cli/main.go:17]: All good, but shouldn't that be in another PR? (it's already in a different commit)

**Review by evrardjp — COMMENTED** [review]
Looking good at first sight.

File: design/api.go [hunk 4]
```diff
@@ -17,13 +17,13 @@ var _ = API("fuseml", func() {
 		Services("runnable", "openapi")
 
 		// List the Hosts and their transport URLs.
-		Host("localhost", func() {
```
Comment (stefannica) [inline design/api.go:20]: I would rather keep this localhost, which is a safer default value (at least for me, when running locally) and override it in the container image entrypoint as 0.0.0.0. It also creates fewer changes overall :)

**Review by stefannica — CHANGES_REQUESTED** [review]

File: cmd/fuseml_core-cli/main.go (hunk 2 above)
Comment (stefannica) [reply cmd/fuseml_core-cli/main.go:102]: I also had cobra in mind, have seen it used in several projects and look easy to use and feature rich. this code is still based on the original generated by goa, but we'll have to improve that at some point.

File: Makefile (hunk 1 above)
Comment (flaviodsr) [reply Makefile:38]: I have added `golint` in addition

File: cmd/fuseml_core-cli/main.go (hunk 2 above)
Comment (flaviodsr) [reply cmd/fuseml_core-cli/main.go]: agreed, at some point we need to improve the CLI. This is just a starting point.

File: cmd/fuseml_core-cli/main.go (hunk 3 above)
Comment (flaviodsr) [reply cmd/fuseml_core-cli/main.go]: For the container to work we need the process to listen in a proper interface. So I just added it to that PR,

File: design/api.go (hunk 4 above)
Comment (flaviodsr) [reply design/api.go:20]: I have updated the code with the addition of the `prod` host that will listen on `0.0.0.0` and kept the default `localhost` as the `dev` host.

Comment (stefannica) [reply design/api.go:20]: thanks

**Review by stefannica — APPROVED** [review]

File: Makefile [hunk 5]
```diff
@@ -1,37 +1,66 @@
...
 
-test:
-	ginkgo ./gen ./pkg
+# Run fuseml_core-cli
+run-fuseml_core-cli: generate lint
+	go run ./cmd/fuseml_core-cli
 
-gen-core:
+# Generate code
+generate:
```
Comment (evrardjp) [inline Makefile:33]: I am not so fond of having this instruction there. It's neat and convenient, but it's so easy that we'll most likely use and abuse this. Generating a new API means breaking existing APi (if I am not mistaken).
I can foresee this becoming a problem in the long run when we'll have at least ONE api user (and because this project is public, I expect that it will arrive soon, at first installation ;) )

So I would rather this be documented than a CLI function. But maybe I am too careful here.

File: cmd/fuseml_core/main.go [hunk 6]
```diff
@@ -119,13 +119,66 @@ func main() {
...
 			} else if u.Port() == "" {
-				u.Host = net.JoinHostPort(u.Host, ":8080")
+				u.Host = net.JoinHostPort(u.Host, "8080")
+			}
+			handleGRPCServer(ctx, u, runnableEndpoints, &wg, errc, logger, *dbgF)
+		}
+
+	case "prod":
+		{
+			addr := "http://0.0.0.0"
```
Comment (evrardjp) [inline cmd/fuseml_core/main.go:129]: Apart from this line, the code below is literally the same code.
Why not extracting this into a function/method which takes the address as argument, and be done with this?

File: design/api.go [hunk 7]
```diff
@@ -17,13 +17,19 @@ var _ = API("fuseml", func() {
...
+		Host("dev", func() {
+			Description("Run the service listening on localhost.")
 			// Transport specific URLs, supported schemes are:
 			// 'http', 'https', 'grpc' and 'grpcs' with the respective default
 			// ports: 80, 443, 8080, 8443.
 			URI("http://localhost:8000")
 			URI("grpc://localhost:8080")
 		})
+
+		Host("prod", func() {
```
Comment (evrardjp) [inline design/api.go:29]: Same as above, instead of using an anonymous function, just use a named function with the right parameter.

File: gen/http/openapi3.yaml [hunk 8]
```diff
@@ -6,6 +6,8 @@ info:
 servers:
 - url: http://localhost:8000
   description: fuseml-core hosts the core services
+- url: http://0.0.0.0
```
Comment (evrardjp) [inline gen/http/openapi3.yaml:9]: Why is this here? Shouldn't the url be flexible? I don't see why there should be two servers there.

**Review by evrardjp — DISMISSED** [review]
A few comments.

File: Dockerfile [hunk 9]
```diff
@@ -0,0 +1,29 @@
+# Build the fuseml_core binary
+FROM golang:1.16 as builder
+
+WORKDIR /workspace
+
+# Copy the Go Modules manifests
+COPY go.mod go.mod
```
[the remark this answers is not in the document]
Comment (flaviodsr) [reply Dockerfile]: done!

File: Dockerfile [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+COPY go.sum go.sum
+
+# cache deps before building and copying source so that we don't need to re-download as much
+# and so that source changes don't invalidate our downloaded layer
+RUN go mod download
+
+# Copy the go source
+COPY gen/ gen/
```
[the remark this answers is not in the document]
Comment (flaviodsr) [reply Dockerfile:14]: Unfortunately for directories `COPY` behaves a bit different. If I try to copy multiple directories it will copy only the contents of the directory instead of the directory itself.

File: Dockerfile [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+COPY cmd/ cmd/
+COPY pkg/ pkg/
+
+# Build
+RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -a -ldflags '-s -w' -o fuseml_core ./cmd/fuseml_core
```
[the remark this answers is not in the document]
Comment (flaviodsr) [reply Dockerfile:19]: Right, for now I would say just leave it like this and if/when we decide to support other archs/OS's make those as variables.

File: Makefile [hunk 12]
```diff
@@ -1,37 +1,66 @@
+# Image URL to use all building/pushing image targets
+IMG ?= fuseml-core:latest
+
+# Get the currently used golang install path (in GOPATH/bin, unless GOBIN is set)
```
[the remark this answers is not in the document]
Comment (flaviodsr) [reply Makefile:4]: You think so?

The idea is that for CI/release we could dynamically use `IMG` env variable to set the container registry/name/tag for pushing it.

File: Makefile [hunk 13]
```diff
@@ -1,37 +1,66 @@
...
+# Build FuseML binaries
+fuseml: generate lint
+	go build ${GO_LDFLAGS} -o bin/fuseml_core ./cmd/fuseml_core
+	go build ${GO_LDFLAGS} -o bin/fuseml_core-cli ./cmd/fuseml_core-cli
 
-build: tidy lint
-	go build ${GO_GCFLAGS} -o bin/fuseml_core ${GO_LDFLAGS} ./cmd/fuseml_core
-	go build ${GO_GCFLAGS} -o bin/fuseml_core-cli ${GO_LDFLAGS} ./cmd/fuseml_core-cli
+# Run fuseml_core
+run-fuseml_core: generate lint
```
[the remark this answers is not in the document]
Comment (flaviodsr) [reply Makefile]: I like it.

File: Makefile (hunk 5 above)
Comment (flaviodsr) [reply Makefile:33]: Sorry but I did not get it. Could you elaborate?

If there is no change on `design` the API generated code won't change. However if you change `design` it will change and I believe having generate here encourages finding a breaking change early.

File: cmd/fuseml_core/main.go (hunk 6 above)
Comment (flaviodsr) [reply cmd/fuseml_core/main.go:129]: This code is auto-generated and we might want to improve it later. For now I prefer to leave it like this as this is not the intent of this PR.

File: design/api.go (hunk 7 above)
Comment (flaviodsr) [reply design/api.go:29]: This comes from the Goa DSL, I tried using a named function and it did not work as it expects a `DSL function`.

File: gen/http/openapi3.yaml (hunk 8 above)
Comment (flaviodsr) [reply gen/http/openapi3.yaml:9]: This code is auto-generated and this comes from the Host definitions on the on `design/api.go`

File: design/api.go (hunk 7 above)
Comment (evrardjp) [reply design/api.go:29]: I can dig there with you if you don't mind.

File: gen/http/openapi3.yaml (hunk 8 above)
Comment (evrardjp) [reply gen/http/openapi3.yaml:9]: I expected this. Still, it makes no sense to have those two :) 
It probably should be linked to an env variable or other.

File: Makefile (hunk 5 above)
Comment (evrardjp) [reply Makefile:33]: Why do we check out the generated code in git, then?

I am fine either way, I just want to understand :)

**Comment by evrardjp** [comment]
@flaviodsr Can you fix conflicts so that I can merge this ?

File: Makefile [hunk 14]
```diff
@@ -1,37 +1,66 @@
+# Image URL to use all building/pushing image targets
+IMG ?= fuseml-core:latest
```
Comment (jsuchome) [inline Makefile:2]: Where should it push, to github registry?

Comment (flaviodsr) [reply Makefile:2]: The way it is it will push to `docker.io/library`.

The idea is to set `IMG` as environment variable and use `docker-build` to build the image with the name provided and later `docker-push` to push it to whatever registry you set on `IMG`. The later would probably be more useful for CI.

**Review by jsuchome — APPROVED** [review]
